polygeist/lib/polygeist/Passes/ConvertPolygeistToLLVM.cpp

```diff
@@ -90,84 +84,131 @@ struct SubIndexOpLowering : public ConvertOpToLLVMPattern<SubIndexOp> {
+           "Expecting the input and output MemRef ranks to be the same");
+    // Note:in MLIRASTConsumer::getMLIRType(), a memref of struct type is
+    // generated only for a struct that has at least one member of SYCL type,
+    // otherwise a llvm pointer type is generated instead of a memref).
+    assert((sycl::isSYCLType(sourceElemType) ||
+            sourceElemType.isa<LLVM::LLVMStructType>() &&
+                any_of(sourceElemType.cast<LLVM::LLVMStructType>().getBody(),
+                       [](const Type &memType) {
+                         return sycl::isSYCLType(memType);
+                       })) &&
+           "the source memref element type should be either a SYCL type, or "
+           "a struct containing at least a member of SYCL type");
+
+    SmallVector<Value, 4> indices;
+    computeIndices(convSourceElemType.cast<LLVM::LLVMStructType>(),
+                   convViewElemType, indices, subViewOp, transformed, rewriter);
+    assert(!indices.empty() && "Expecting a least one index");
+
+    // According to MLIRScanner::InitializeValueByInitListExpr() in
+    // clang-mlir.cc, when a memref element type is a struct type, the return
+    // type of a polygeist.subindex operation should be a memref of the
+    // element type of the struct.
+    auto elemPtrTy = LLVM::LLVMPointerType::get(convViewElemType);
+    auto gep = rewriter.create<LLVM::GEPOp>(loc, elemPtrTy, prev, indices);
+    auto memRefDesc = createMemRefDescriptor(loc, viewMemRefType, gep, gep,
+                                             sizes, strides, rewriter);
+    LLVM_DEBUG(llvm::dbgs() << "gep: " << *gep << "\n");
+
+    rewriter.replaceOp(subViewOp, {memRefDesc});
     return success();
   }
+
+private:
+  // Compute the indices of the GEP operation we lower the SubIndexOp to.
+  // The indices are computed based on:
+  //   a) the (converted) source element type, and
+  //   b) the (converted) result element type that is requested
+  // Examples:
+  //  - src ty: ptr<struct<array<1xi64>>>, res ty: ptr<i64>
+  //      -> idxs = [0, 0, SubIndexOp's index]
+  //  - src ty: ptr<struct<array<1xi64>>>, res ty: ptr<array<1xi64>>
+  //      -> idxs = [0, SubIndexOp's index]
+  //
+  // Note: the (converted) result type that is requested is deemed illegal
+  // unless it is one of the source member types. For example assume:
+  //   - src ty: ptr<struct<array<1xi64>,i32>>
+  //   - res ty: ptr<i64>
+  // This is illegal because res ty can only be either ptr<i32> or
+  // ptr<array<1xi64>>
+  void computeIndices(const LLVM::LLVMStructType &srcElemType,
+                      const Type &resElemType, SmallVectorImpl<Value> &indices,
+                      SubIndexOp op, OpAdaptor transformed,
+                      ConversionPatternRewriter &rewriter) const {
+    assert(indices.empty() && "Expecting an empty vector");
+
+    ArrayRef<Type> memTypes = srcElemType.getBody();
+    unsigned numMembers = memTypes.size();
+    assert((numMembers == 1 ||
```

Comment:
We may want to check `numMembers == 1` recursively to allow cases like:
```
  //  - src ty: ptr<struct<struct<array<1xi64>>>>, res ty: ptr<i64>
  //      -> idxs = [0, 0, 0, SubIndexOp's index]
```

Reply:
This should work already because the TypeSwitch checks whether the member of the struct is itself a struct with a single member at lines 194-197

Reply:
good point, do we want to assert for cases like:
```
  //   - src ty: ptr<struct<struct<array<1xi64>,i32>>>
  //   - res ty: ptr<i8>
```

Reply:
yes we assert for those cases.

Reply:
right, at line 205.